kerastuner/engine/oracle.py

```diff
@@ -256,22 +269,62 @@ def remaining_trials(self):
             return None
 
     def get_state(self):
-        state = {}
-        state['trials'] = {trial_id: trial.get_state()
-                           for trial_id, trial in self.trials.items()}
+        # `self.trials` are saved in their own, Oracle-agnostic files.
         # Just save the IDs for ongoing trials, since these are in `trials`.
+        state = {}
         state['ongoing_trials'] = {
             tuner_id: trial.trial_id
             for tuner_id, trial in self.ongoing_trials.items()}
+        # Hyperparameters are part of the state because they can be added to
+        # during the course of the search.
+        state['hyperparameters'] = self.hyperparameters.get_config()
         return state
 
     def set_state(self, state):
-        self.trials = {
-            trial_id: trial_lib.Trial.from_state(trial_config)
-            for trial_id, trial_config in state['trials'].items()}
+        # `self.trials` are saved in their own, Oracle-agnostic files.
         self.ongoing_trials = {
             tuner_id: self.trials[trial_id]
             for tuner_id, trial_id in state['ongoing_trials'].items()}
+        self.hyperparameters = hp_module.HyperParameters.from_config(
+            state['hyperparameters'])
+
+    def set_project_dir(self, directory, project_name, reload=True):
```

Comment:
If saving / reloading is controlled by the Tuner, then the Oracle should preferably not own a project_dir. Instead, it should expose `reload(path)` and `save(path)` methods, and these methods should be called by the Tuner. It seems like we never want a situation where Oracle and Tuner are saved in different places.

Reply:
The issue is that for fault-tolerance, the Oracle needs to be able to save trials in methods like `create_trial`, `update_trial`, `end_trial`, and needs to be able to save itself in `create_trial` and `end_trial`.

Agreed that the project_dir should always be the same as the Tuner's. Maybe we could just make this method private? It is only called in the __init__ of BaseTuner (and in tests), and users wouldn't be expected to call it

Reply:
Let's see:
- Oracle needs a directory
- It needs to be set by the Tuner, so there needs to be a method to set
- We should only have 1 way to set the directory for the oracle, so we can't also have a constructor argument
- We don't want users to set it manually and it should always be in sync with the Tuner's dir, so the method to set it must be private
- But if it can only be set via a private method, methods like `save` and `reload` don't make sense as user-facing methods

Maybe we should set all new methods as private?

Reply:
SGTM, done. `_save`, `_reload`, `_project_dir` and `_set_project_dir` made private

Reply:
Actually, can we keep `save` and `reload` public? They already had implementations via the inheritance from `Stateful`, and the implementation added here relies on the super version there. Also this way the contract of how `get_state` / `set_state` and `save` / `reload` interact is more clear. WDYT?

Reply:
SGTM